### PR TITLE
Disable batch transfer in DP mode

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

There is no need to transfer the batch to the device in DP mode since it's handled internally. Else what's happening now is let say a batch with batch_size 32 is there. Then first it's transferred to gpu:0, passed to the model wrapped around DataParallel and again transferred to separate devices after splitting. So the first transfer is unnecessary.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -66,9 +66,9 @@ def setup(self, trainer, model: LightningModule) -> None:
             trainer: the trainer instance to connect to
             model: the model to train
         """
-        self.connect_training_type_plugin(self.training_type_plugin, model)
+        self.connect_training_type_plugin(model)
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:69]: The changes in this file are not relevant for the PR, are they?

File: pytorch_lightning/accelerators/gpu.py [hunk 2]
```diff
@@ -38,3 +40,10 @@ def set_nvidia_flags():
         all_gpu_ids = ",".join([str(x) for x in range(torch.cuda.device_count())])
         devices = os.getenv("CUDA_VISIBLE_DEVICES", all_gpu_ids)
         _log.info(f"LOCAL_RANK: {os.getenv('LOCAL_RANK', 0)} - CUDA_VISIBLE_DEVICES: [{devices}]")
+
+    def to_device(self, batch: Any) -> Any:
```
Comment (justusschock) [inline pytorch_lightning/accelerators/gpu.py:44]: Actually I'm afraid we shouldn't add this right now, since that way the user cannot device custom batch-transfers for custom types. This is already true in the current case, but we should rather enable this instead of introducing more blockers.

The found solution can then be applied to this fix as well :)

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 3]
```diff
@@ -81,7 +81,8 @@ def attach_data(self, model, train_dataloader, val_dataloaders, datamodule):
 
         # set up the passed in dataloaders (if needed)
         self.attach_dataloaders(model, train_dataloader, val_dataloaders)
-        self.attach_datamodule(model, datamodule, 'fit')
+        self.attach_datamodule(model, datamodule)
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/data_connector.py:84]: Those parts also aren't relevant for this PR, right?

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -915,8 +915,9 @@ def test(
                 'You cannot pass test_dataloaders to trainer.test if you supply a datamodule'
             )
 
-        # Attach datamodule to get setup/prepare_data added to model before the call to it below
-        self.data_connector.attach_datamodule(model or self.lightning_module, datamodule, 'test')
+        if datamodule is not None:
```
Comment (justusschock) [inline pytorch_lightning/trainer/trainer.py:918]: also not relevant here...

File: tests/accelerators/test_dp.py [hunk 5]
```diff
@@ -95,3 +98,53 @@ def test_dp_test(tmpdir):
...
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+def test_dp_raise_exception_with_batch_transfer_hooks(tmpdir):
+    """
+    Test that an exception is raised when overriding batch_transfer_hooks in DP model.
+    """
+
+    class CustomModel(BoringModel):
+
+        def transfer_batch_to_device(self, batch, device):
+            batch = batch.cuda()
```
Comment (justusschock) [inline tests/accelerators/test_dp.py:112]:
```suggestion
            batch = batch.to(device)
```

File: tests/accelerators/test_dp.py [hunk 6]
```diff
@@ -95,3 +98,53 @@ def test_dp_test(tmpdir):
...
+    model = CustomModel()
+
+    with pytest.raises(MisconfigurationException, match=r'Overriding `transfer_batch_to_device` is not .* in DP'):
+        trainer.fit(model)
+
+    # Override on_before_batch_transfer hook only
+    trainer = Trainer(**trainer_options)
+    model = CustomModel()
+    model.transfer_batch_to_device = BoringModel().transfer_batch_to_device
+    model.on_before_batch_transfer = model.transform_hook
```
Comment (justusschock) [inline tests/accelerators/test_dp.py:138]: Can you avoid patching by just instantiating boring model or proper subclassing?

File: tests/accelerators/test_dp.py [hunk 7]
```diff
@@ -95,3 +98,53 @@ def test_dp_test(tmpdir):
...
+
+    with pytest.raises(MisconfigurationException, match=r'Overriding `on_before_batch_transfer` is not .* in DP'):
+        trainer.fit(model)
+
+    # Override on_after_batch_transfer hook only
+    trainer = Trainer(**trainer_options)
+    model = CustomModel()
+    model.transfer_batch_to_device = BoringModel().transfer_batch_to_device
+    model.on_after_batch_transfer = model.transform_hook
```
Comment (justusschock) [inline tests/accelerators/test_dp.py:147]: also avoid patching

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py:69]: already reverted locally. was checking something.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:84]: already reverted.

File: pytorch_lightning/accelerators/gpu.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/gpu.py:44]: even if they have a custom batch, will DataParallel be able to split that batch across each gpu??

Comment (justusschock) [reply pytorch_lightning/accelerators/gpu.py:44]: That's what I meant :) Yeah. currently this is an issue for us. 
If we want to proceed with this, maybe just mention it in a comment and the docs?

Comment (rohitgr7) [reply pytorch_lightning/accelerators/gpu.py:44]: done.

File: pytorch_lightning/accelerators/gpu.py [hunk 8]
```diff
@@ -38,3 +40,11 @@ def set_nvidia_flags():
...
+        # no need to transfer batch to device in DP mode
+        # TODO: Add supported to allow batch transfer to device in Lightning for DP mode.
```
Comment (carmocca) [inline pytorch_lightning/accelerators/gpu.py:46]:
```suggestion
        # TODO: Add support to allow batch transfer to device in Lightning for DP mode.
```

File: pytorch_lightning/core/hooks.py [hunk 9]
```diff
@@ -609,6 +610,10 @@ def transfer_batch_to_device(self, batch, device):
                     batch = super().transfer_batch_to_device(data, device)
                 return batch
 
+        Raises:
+            MisconfigurationException:
+                If using data-parallel(``accelerator='dp'``) in ``Trainer``.
```
Comment (carmocca) [inline pytorch_lightning/core/hooks.py:615]:
```suggestion
                If using data-parallel, ``Trainer(accelerator='dp')``.
```

File: pytorch_lightning/core/hooks.py [hunk 10]
```diff
@@ -620,10 +625,11 @@ def on_before_batch_transfer(self, batch, dataloader_idx):
         """
         Override to alter or apply batch augmentations to your batch before it is transferred to the device.
 
-        .. warning:: dataloader_idx always returns 0, and will be updated to support the true idx in the future.
+        .. warning:: ``dataloader_idx`` always returns 0, and will be updated to support the true ``idx`` in the future.
```
Comment (carmocca) [inline pytorch_lightning/core/hooks.py:628]:
```suggestion
        .. warning:: ``dataloader_idx`` always returns 0, and will be updated to support the true index in the future.
```

File: pytorch_lightning/core/hooks.py [hunk 11]
```diff
@@ -638,6 +644,10 @@ def on_before_batch_transfer(self, batch, dataloader_idx):
                 batch['x'] = transforms(batch['x'])
                 return batch
 
+        Raises:
+            MisconfigurationException:
+                If using data-parallel(``accelerator='dp'``) in ``Trainer``.
```
Comment (carmocca) [inline pytorch_lightning/core/hooks.py:649]:
```suggestion
                If using data-parallel, ``Trainer(accelerator='dp')``.
```

File: pytorch_lightning/core/hooks.py [hunk 12]
```diff
@@ -666,6 +677,10 @@ def on_after_batch_transfer(self, batch, dataloader_idx):
                 batch['x'] = gpu_transforms(batch['x'])
                 return batch
 
+        Raises:
+            MisconfigurationException:
+                If using data-parallel(``accelerator='dp'``) in ``Trainer``.
```
Comment (carmocca) [inline pytorch_lightning/core/hooks.py:682]:
```suggestion
                If using data-parallel, ``Trainer(accelerator='dp')``.
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 13]
```diff
@@ -90,6 +91,14 @@ def __enforce_datamodule_dataloader_override(self, train_dataloader, val_dataloa
                 'You cannot pass train_dataloader or val_dataloaders to trainer.fit if you supply a datamodule'
             )
 
+    def _validate_data_hooks(self, model):
+        # Raise Misconfiguration exception since these hooks are not supported in DP mode
+        # TODO: Remove this blocker once batch transfer to device is integrated in Lightning for DP mode.
+        batch_transfer_hooks = ['on_before_batch_transfer', 'transfer_batch_to_device', 'on_after_batch_transfer']
+        for hook in batch_transfer_hooks:
+            if self.trainer.accelerator_connector.use_dp and is_overridden(hook, model):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:99]: I think this works?

```suggestion
            if self.trainer.use_dp and is_overridden(hook, model):
```

File: tests/accelerators/test_dp.py [hunk 14]
```diff
@@ -95,3 +98,54 @@ def test_dp_test(tmpdir):
     new_weights = model.c_d1.weight.clone().detach().cpu()
 
     assert torch.all(torch.eq(old_weights, new_weights))
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
```
Comment (carmocca) [inline tests/accelerators/test_dp.py:103]: We won't be really training for this tests right? Can you mock the device_count?

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 13 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:99]: deprecated: https://github.com/PyTorchLightning/pytorch-lightning/blob/ae6ce17598cde260f980079b5b3a4274c19eb7a1/pytorch_lightning/trainer/deprecated_api.py#L73-L75

File: tests/accelerators/test_dp.py (hunk 14 above)
Comment (rohitgr7) [reply tests/accelerators/test_dp.py:103]: It needs the GPUs to initialize the DP plugin else I think it stays on CPU.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 15]
```diff
@@ -115,29 +124,23 @@ def attach_dataloaders(
...
             # Override loader hooks
-            if is_overridden('train_dataloader', datamodule):
-                model.train_dataloader = datamodule.train_dataloader
-            if is_overridden('val_dataloader', datamodule):
-                model.val_dataloader = datamodule.val_dataloader
-            if is_overridden('test_dataloader', datamodule):
-                model.test_dataloader = datamodule.test_dataloader
-            if is_overridden('predict_dataloader', datamodule):
-                model.predict_dataloader = datamodule.predict_dataloader
+            dl_methods = ['train_dataloader', 'val_dataloader', 'test_dataloader', 'predict_dataloader']
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/data_connector.py:134]:
```suggestion
            dl_methods = ('train_dataloader', 'val_dataloader', 'test_dataloader', 'predict_dataloader')
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 16]
```diff
@@ -115,29 +124,23 @@ def attach_dataloaders(
...
+                    setattr(model, method, getattr(datamodule, method))
 
             # Override data transfer hooks if dataset-specific to_device logic has been defined in datamodule
-            if is_overridden('on_before_batch_transfer', datamodule):
-                model.on_before_batch_transfer = datamodule.on_before_batch_transfer
-            if is_overridden('transfer_batch_to_device', datamodule):
-                model.transfer_batch_to_device = datamodule.transfer_batch_to_device
-            if is_overridden('on_after_batch_transfer', datamodule):
-                model.on_after_batch_transfer = datamodule.on_after_batch_transfer
+            batch_transfer_hooks = ['on_before_batch_transfer', 'transfer_batch_to_device', 'on_after_batch_transfer']
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/data_connector.py:140]:
```suggestion
            batch_transfer_hooks = ('on_before_batch_transfer', 'transfer_batch_to_device', 'on_after_batch_transfer')
```

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 15 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:134]: Nice cleaning !

**Review by tchaton — APPROVED** [review]
Overall, looks good !

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 17]
```diff
@@ -90,6 +91,14 @@ def __enforce_datamodule_dataloader_override(self, train_dataloader, val_dataloa
                 'You cannot pass train_dataloader or val_dataloaders to trainer.fit if you supply a datamodule'
             )
 
+    def _validate_data_hooks(self, model):
+        # Raise Misconfiguration exception since these hooks are not supported in DP mode
+        # TODO: Remove this blocker once batch transfer to device is integrated in Lightning for DP mode.
+        batch_transfer_hooks = ['on_before_batch_transfer', 'transfer_batch_to_device', 'on_after_batch_transfer']
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/data_connector.py:97]:
```suggestion
        batch_transfer_hooks = ('on_before_batch_transfer', 'transfer_batch_to_device', 'on_after_batch_transfer')
```

File: tests/accelerators/test_dp.py (hunk 14 above)
Comment (carmocca) [reply tests/accelerators/test_dp.py]: Seems to work: https://github.com/PyTorchLightning/pytorch-lightning/pull/6098/commits/0034c5c6e129f274cf0dbb35ba4e455db56eac67

**Review by carmocca — APPROVED** [review]

File: tests/accelerators/test_dp.py (hunk 14 above)
Comment (rohitgr7) [reply tests/accelerators/test_dp.py]: oh okay.. thanks :v:

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
